plugins/suspendmanager.cpp

```diff
@@ -0,0 +1,822 @@
+        }
+        return false;
+    }
+
+    static bool tileHasSupportFloor(coord pos) {
+        auto tile_type = Maps::getTileType(pos);
+        if (tile_type) {
+            auto shape = tileShape(*tile_type);
+            return shape == df::enums::tiletype_shape::NONE ? false : shape_floor_support[shape];
+        }
+        return false;
+    }
+
+    static bool tileHasSupportBuilding(coord pos) {
+        auto building = Buildings::findAtTile(pos);
+        if (building)
+            return building->getType() == df::building_type::Support && building->flags.bits.exists;
+        return false;
+    }
+
+
+    static bool isImpassable(df::building* building) {
+        auto type = building->getType();
+        if (type == building_type::Construction) {
+            return construction_impassible[building->getSubtype()];
+        }
+        else return building_impassible[type];
+    }
+
+    static bool hasWalkableNeighbor(coord pos) {
+        return std::ranges::any_of(neighbors | transform(around(pos)), walkable);
+    }
+
+
+    static df::building* plannedImpassibleAt(coord pos) {
+        auto building = Buildings::findAtTile(pos);
+        if (!building)
+            return nullptr;
+        if (!building->flags.bits.exists && isImpassable(building))
+            return building;
+        return nullptr;
+    }
+
+    static bool isBuildingPlanJob (df::job* job) {
+        // TODO check that this it correct
+        auto building = Job::getHolder(job);
+        return building && building->mat_type == -1;
+    }
+
+    static int riskOfStuckConstructionAt(coord pos) {
+        auto risk = 0;
+        for (auto npos : neighbors | transform(around(pos))) {
+            if (!isSuitableAccess(npos)) { // original used walkable
+                ++risk; // one access blocked, increase danger
+            } else if (!plannedImpassibleAt(npos)) {
+                // walkable neighbour with no plan to build a wall, no danger
+                return -1;
+            }
+        }
+        return risk;
+    }
+
+
+
+    // return true if this job is at risk of blocking another one
+    static bool riskBlocking(color_ostream &out, df::job* job) {
+        if (job->job_type != job_type::ConstructBuilding)
+            return false;
+        TRACE(cycle,out).print("risk blocking: check construction job %d\n", job->id);
+
+        auto building = Job::getHolder(job);
+        if (!building || !isImpassable(building))
+            return false; // not building a blocking construction, no risk
+
+        coord pos(building->centerx,building->centery,building->z);
+        if (!isSuitableAccess(pos))
+            return false; // construction is on a non-walkable tile, can't block
+
+        auto risk = riskOfStuckConstructionAt(pos);
+        TRACE(cycle,out).print("  risk is %d\n", risk);
+
+        // TOTHINK: on a large grid, this will compute the same risk up to 5 times
```

Comment:
would have to test to see if it matters -- in previous tests (in other plugins), maintaining a hashset to reduce duplication was slower than just doing the work multiple times